presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/KDBTree.java

```diff
@@ -0,0 +1,344 @@
+        addLeafNodes(rootNode, leafNodes);
+        return leafNodes.build();
+    }
+
+    private static void addLeafNodes(Node node, ImmutableMap.Builder<Integer, Rectangle> leafNodes)
+    {
+        if (node.leafId.isPresent()) {
+            leafNodes.put(node.leafId.getAsInt(), node.extent);
+        }
+        else {
+            addLeafNodes(node.left.get(), leafNodes);
+            addLeafNodes(node.right.get(), leafNodes);
+        }
+    }
+
+    public Map<Integer, Rectangle> findIntersectingLeafNodes(Rectangle envelope)
+    {
+        ImmutableMap.Builder<Integer, Rectangle> leafNodes = ImmutableMap.builder();
+        addIntersectingLeafNodes(envelope, rootNode, leafNodes);
+        return leafNodes.build();
+    }
+
+    private static void addIntersectingLeafNodes(Rectangle envelope, Node node, ImmutableMap.Builder<Integer, Rectangle> leafNodes)
+    {
+        if (node.extent.disjoint(envelope)) {
+            return;
+        }
+
+        if (node.leafId.isPresent()) {
+            leafNodes.put(node.leafId.getAsInt(), node.extent);
+        }
+        else {
+            addIntersectingLeafNodes(envelope, node.left.get(), leafNodes);
+            addIntersectingLeafNodes(envelope, node.right.get(), leafNodes);
+        }
+    }
+
+    private static final Comparator<Rectangle> BY_X = (first, second) -> ComparisonChain.start()
+            .compare(first.getXMin(), second.getXMin())
+            .compare(first.getYMin(), second.getYMin())
+            .result();
+
+    private static final Comparator<Rectangle> BY_Y = (first, second) -> ComparisonChain.start()
+            .compare(first.getYMin(), second.getYMin())
+            .compare(first.getXMin(), second.getXMin())
+            .result();
+
+    private static final class LeafIdAllocator
+    {
+        private int nextId;
+
+        public int next()
+        {
+            return nextId++;
+        }
+    }
+
+    public static KDBTree buildKDBTree(int maxItemsPerNode, int maxLevels, Rectangle extent, List<Rectangle> items)
+    {
+        checkArgument(maxItemsPerNode > 0, "maxItemsPerNode must be > 0");
+        checkArgument(maxLevels > 0, "maxLevels must be > 0");
+        requireNonNull(extent, "extent is null");
+        requireNonNull(items, "items is null");
+        return new KDBTree(buildKDBTreeNode(maxItemsPerNode, maxLevels, 0, extent, items, new LeafIdAllocator()));
+    }
+
+    private static Node buildKDBTreeNode(int maxItemsPerNode, int maxLevels, int level, Rectangle extent, List<Rectangle> items, LeafIdAllocator leafIdAllocator)
+    {
+        checkArgument(maxItemsPerNode > 0, "maxItemsPerNode must be > 0");
+        checkArgument(maxLevels > 0, "maxLevels must be > 0");
+        checkArgument(level >= 0, "level must be >= 0");
+        requireNonNull(extent, "extent is null");
+        requireNonNull(items, "items is null");
+
+        if (items.size() <= maxItemsPerNode || level >= maxLevels) {
```

Comment:
No.

Reply:
Then use an equality check and a verify() to check the invariant. Using >= can mask subtle bugs if levels is not supposed to be greater than maxLevels.

Reply:
@martint Using `verify` would be inconsistent with using `checkArgument` for `level >= 0` check. At the same time, if I use `checkArgument` it looks quite strange:

```
        checkArgument(maxItemsPerNode > 0, "maxItemsPerNode must be > 0");
        checkArgument(level >= 0, "level must be >= 0");
        checkArgument(level <= 10_000, "level must be <= 10_000");
        requireNonNull(extent, "extent is null");
        requireNonNull(items, "items is null");

        if (items.size() <= maxItemsPerNode || level == 10_000) {
            return newLeaf(extent, leafIdAllocator.next());
        }
```

What do you think?

Reply:
Ah, yes, since they are arguments it should use checkArgument. I don't find it strange, btw. You may want to introduce a constant to capture the 10_000 magic value, and make the code a little more future-proof.